pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -202,34 +202,46 @@ func (m *ManagerImpl) Stop() error {
 }
 
 func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {
+	existingDevs := make(map[string]pluginapi.Device)
+	m.mutex.Lock()
+	old, ok := m.endpoints[r.ResourceName]
+	if ok && old != nil {
+		// Pass devices of previous endpoint into re-registered one,
+		// to avoid potential orphaned devices upon re-registration
+		existingDevs = old.devices
+	}
+	m.mutex.Unlock()
+
 	socketPath := filepath.Join(m.socketdir, r.Endpoint)
-	e, err := newEndpoint(socketPath, r.ResourceName, m.callback)
+	e, err := newEndpoint(socketPath, r.ResourceName, existingDevs, m.callback)
 	if err != nil {
 		glog.Errorf("Failed to dial device plugin with request %v: %v", r, err)
 		return
 	}
 
-	stream, err := e.list()
-	if err != nil {
-		glog.Errorf("Failed to List devices for plugin %v: %v", r.ResourceName, err)
+	m.mutex.Lock()
+	// Check for potential re-registration during the initialization of new endpoint,
+	// and skip updating if re-registration happens.
+	// TODO: simplify the part once we have a better way to handle registered devices
+	ext := m.endpoints[r.ResourceName]
+	if ext != old {
+		glog.Warningf("Some other endpoint %v is added while endpoint %v is initialized", ext, e)
+		m.mutex.Unlock()
 		e.stop()
 		return
 	}
```

Comment:
this looks ugly to me ... it could have been implemented differently, using a single, short lock while avoiding reregistration:

```
func (m *ManagerImpl) addEndpoint(r *pluginapi.RegisterRequest) {

    // The call involves an initial dial up which may take time to conclude
    e, err := newEndpoint(m, r)
    if err != nil {
        glog.Errorf("Failed to initialize endpoint %v: %v", r.Endpoint, err)
        return
    }

    // Now we lock the endpoints for mutation
    registered := false
    m.mutex.Lock()
    curr, ok := m.endpoints[r.ResourceName]
    if ok & curr != nil {
       registered = true
    } else {
       m.endpoints[r.ResourceName] = e
    }
    m.mutex.Unlock()

    if registered {
       e.stop()
       e = curr
    } else {
       curr.stop()
    }

    go func() {
       e.run()
       ...
    }()
}
```

In pkg/kubelet/cm/deviceplugin/endpoint.go:

```
func newEndpoint(m *Manager, resourceName string, r *pluginapi.RegisterRequest) (*endpoint, error) {
    socketPath := filepath.Join(m.SocketPath(), r.Endpoint)
    client, c, err := dial(socketPath)
    if err != nil {
        return nil, err
    }
    return &endpoint{
       client: client,
       clientConn: c,
       socketPath: socketPath,
       resourceName: resourceName,
       callback: m.callback,
    }  
}
```

Reply:
Well, the issue of re-registration is raised by the original purpose of passing existing devs into `newEndpoint` , plz take a look at #55179 and comments above

Reply:
well, I will check that PR review history. At the same time, I would suggest you to try review the logic I suggested above and see if it help simplify the whole function. It would also be a help to me so I know if I'm misunderstanding something.